bindings/rust/s2n-tls-tokio/src/lib.rs

```diff
@@ -147,6 +149,70 @@ where
             stream.poll_write(async_context, src)
         })
     }
+
+    pub fn get_ref(&self) -> &Connection {
+        &self.conn
+    }
+
+    pub fn get_mut(&mut self) -> &mut Connection {
+        &mut self.conn
+    }
+}
+
+impl<S> AsyncRead for TlsStream<S>
+where
+    S: AsyncRead + AsyncWrite + Unpin,
+{
+    fn poll_read(
+        self: Pin<&mut Self>,
+        ctx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        self.get_mut()
+            .with_io(|mut context| {
+                context.conn.set_waker(Some(ctx.waker()))?;
+                context.conn.recv(buf.initialize_unfilled()).map_ok(|size| {
+                    buf.advance(size);
+                })
+            })
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e))
+    }
```

Comment:
I'm not sure about this error handling. This method must return an io error, and there are defined "Kinds" of io error: https://doc.rust-lang.org/std/io/enum.ErrorKind.html I suppose I could try to map some s2n-tls errors to io errors, but I question how useful that mapping would be.

Another option might be to store any io errors produced by the underlying stream during the send/recv callbacks, and retrieve those errors here?